ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala

```diff
@@ -963,6 +966,93 @@ private class PostgresLedgerDao(
     }
   }
 
+  // Note: package upload is idempotent
+  private val SQL_INSERT_PACKAGE =
+    """insert into packages(package_id, submission_id, source_description, size, known_since, ledger_offset, package)
+          |select {package_id}, {submission_id}, {source_description}, {size}, {known_since}, ledger_end, {package}
+          |from parameters
+          |on conflict (package_id)
+          |do nothing
+          |""".stripMargin
+
+  private val SQL_SELECT_PACKAGES =
+    SQL("""select package_id, source_description, known_since, size
+          |from packages
+          |""".stripMargin)
+
+  private val SQL_SELECT_PACKAGE =
+    SQL("""select package
+          |from packages
+          |where package_id = {package_id}
+          |""".stripMargin)
+
+  case class ParsedPackageData(
+      packageId: String,
+      sourceDescription: Option[String],
+      size: Long,
+      knownSince: Date)
+
+  private val PackageDataParser: RowParser[ParsedPackageData] =
+    Macro.parser[ParsedPackageData](
+      "package_id",
+      "source_description",
+      "size",
+      "known_since"
+    )
+
+  override def listLfPackages: Future[Map[PackageId, PackageDetails]] =
+    dbDispatcher.executeSql { implicit conn =>
+      SQL_SELECT_PACKAGES
+        .as(PackageDataParser.*)
+        .map(
+          d =>
+            PackageId.assertFromString(d.packageId) -> PackageDetails(
+              d.size,
+              d.knownSince.toInstant,
+              d.sourceDescription))
+        .toMap
+    }
+
+  override def getLfArchive(packageId: PackageId): Future[Option[Archive]] =
+    dbDispatcher.executeSql { implicit conn =>
+      SQL_SELECT_PACKAGE
+        .on(
+          "package_id" -> packageId
+        )
+        .as[Option[Array[Byte]]](SqlParser.byteArray("package").singleOpt)
+        .map(data => Archive.parseFrom(Decode.damlLfCodedInputStreamFromBytes(data)))
+    }
+
+  override def uploadLfPackages(
+      packages: List[(Archive, PackageDetails)]): Future[UploadPackagesResult] = {
+    dbDispatcher.executeSql { implicit conn =>
+      Try {
+        val submissionId = UUID.randomUUID().toString
+        val namedPackageParams = packages
+          .map(
+            p =>
+              Seq[NamedParameter](
+                "package_id" -> p._1.getHash,
+                "submission_id" -> submissionId,
+                "source_description" -> p._2.sourceDescription,
+                "size" -> p._2.size,
+                "known_since" -> p._2.knownSince,
+                "package" -> p._1.toByteArray
+            )
+          )
+        executeBatchSql(
+          SQL_INSERT_PACKAGE,
+          namedPackageParams
+        )
+        UploadPackagesResult.Ok
+      }.recover {
+        case NonFatal(e) =>
+          conn.rollback()
```

Comment:
returning an `UploadPackagesResult` should not be the concern of this class. That way you can avoid the `Try/rollback` logic, and just call `transform` on the resulting `Future` to do the mapping to `UploadPackagesResult`

Reply:
I see this is not necessary for the packages. It was done for party management, to differentiate duplicate parties from other database errors. Would you have party allocation method return `Future[Unit]`, and map all failures to `PartyAllocationResult.AlreadyExists`? Or create a custom Ok/Duplicate/OtherError sum type?

Reply:
Oh I see. In that case I would probably follow that pattern like here: https://github.com/digital-asset/daml/blob/c23af0eaa04a63bdda6e58c964a1b824448c7f96/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala#L553